### PR TITLE
Surface.fblits()

### PR DESCRIPTION
**Pull request by itzpr3d4t0r**
Adds the `Surface.fblits(blit_sequence, blit_flag)` function. A faster version of `Surface.blits()` that leverages:
- A faster calling convention (FASTCALL)
- Faster sequence looping
- A single blend flag instead of N flags to process individually
- Always draws the entire surface, so no surface slicing through rects
- Doesn't return a list of rects

The function call is also simpler, as with `blits` if you want to blit with a BLEND_ADD flag and blit the entire images you'd write something like this:
```Python
surf.blits([(obj.img, obj.pos, None, pygame.BLEND_ADD) for obj in to_draw])
```
Whereas with fblits you could do
```Python
surf.fblits([(obj.img, obj.pos) for obj in to_draw], pygame.BLEND_ADD)  # btw you can even pass a Rect as pos as in blits
```
Basically fblits takes in a single blend flag that's then applied to all images in the sequence and draws all images entirely.

### CODE REVIEW

**Comment by MyreMylar** [comment]
looks like there is an actual linting fail here on `blit_test.py`. Needs a quick `black test/blit_test.py`.

**Comment by itzpr3d4t0r** [comment]
> looks like there is an actual linting fail here on `blit_test.py`. Needs a quick `black test/blit_test.py`.

done

File: test/blit_test.py [hunk 1]
```diff
@@ -130,6 +128,73 @@ def blits(blit_list):
...
+        for i in range(NUM_SURFS):
+            color = (i * 1, i * 1, i * 1)
+            self.assertEqual(dst.get_at((i * 10, 0)), color)
+            self.assertEqual(dst.get_at(((i * 10) + 5, 5)), color)
+
+        t0 = time()
+        results = dst.fblits(blit_list, 0)
+        t1 = time()
+        if PRINT_TIMING:
+            print(f"Surface.fblits doreturn=0: {t1 - t0}")
```
Comment (Matiiss) [inline test/blit_test.py:189]: "doreturn=0"?

**Review by Matiiss — CHANGES_REQUESTED** [review]
Lovely idea, main function looks alright.

Mainly some minor inconsistencies in docs and names, also a possible issue with positional-only argument syntax in type stubs.

File: buildconfig/stubs/pygame/surface.pyi [hunk 2]
```diff
@@ -70,6 +70,11 @@ class Surface:
         ],
         doreturn: Union[int, bool] = 1,
     ) -> Union[List[Rect], None]: ...
+    def fblits(
+            self,
+            blit_sequence: Sequence[Tuple[Surface, Union[Coordinate, RectValue]]],
+            blend_flags: int = 0, /
```
[the remark this answers is not in the document]
Comment (Matiiss) [reply buildconfig/stubs/pygame/surface.pyi:76]: Tested another branch with this syntax on 3.7.9 with mypy and it found no issues, the original issue I got this from was this: https://github.com/pygame/pygame/issues/3572

[the remark this answers is not in the document]
Comment (Matiiss) [reply buildconfig/stubs/pygame/surface.pyi:76]: So, apparently I made some mistake while testing and this is in fact an issue still.

File: docs/reST/ref/surface.rst [hunk 3]
```diff
@@ -154,6 +154,28 @@
 
       .. ## Surface.blits ##
 
+   .. method:: fblits
+
+      | :sl:`draw many images onto another, all with the full rect and same blend_flag`
+      | :sg:`fblits(blit_sequence=((source, dest), ...), blend_flags=0) -> None`
+
+      Draws many surfaces onto this Surface. It takes a sequence of tuples (source, dest) as input,
+      and a blend_flags parameter that applies to every surface being drawn. All surfaces are fully drawn.
```
[the remark this answers is not in the document]
Comment (itzpr3d4t0r) [reply docs/reST/ref/surface.rst:163]: basically in blits you can also pass a rect as "area" that defines which section of the original surface to draw. This isn't possible in fblits.

File: buildconfig/stubs/pygame/surface.pyi (hunk 2 above)
[the remark this answers is not in the document]
Comment (itzpr3d4t0r) [reply buildconfig/stubs/pygame/surface.pyi:76]: no prob i removed that.

**Comment by itzpr3d4t0r** [comment]
Don't merge this PR yet pls

**Comment by MyreMylar** [comment]
> Don't merge this PR yet pls

I will not then :)

**Comment by itzpr3d4t0r** [comment]
> > Don't merge this PR yet pls
> 
> I will not then :)

I'm doing refactors, adding comments and a change that should speed up gens but should be checked.

File: docs/reST/ref/surface.rst [hunk 4]
```diff
@@ -154,6 +154,30 @@
...
       .. ## Surface.blits ##
 
+   .. method:: fblits
+
+      | :sl:`draw many surfaces onto the calling surface at their corresponding location and the same special_flags`
+      | :sg:`fblits(blit_sequence=((source, dest), ...), special_flags=0) -> None`
+
+      This method takes a sequence of tuples (source, dest) as input, where source is a Surface
+      object and dest is its destination position on this Surface. It draws each source surface
+      fully(meaning that unlike `blit()` you cannot pass an "area" parameter to represent
```
Comment (Matiiss) [inline docs/reST/ref/surface.rst:164]: there should be a space between "fully" and "("

File: docs/reST/ref/surface.rst [hunk 5]
```diff
@@ -154,6 +154,30 @@
 
       .. ## Surface.blits ##
 
+   .. method:: fblits
+
+      | :sl:`draw many surfaces onto the calling surface at their corresponding location and the same special_flags`
+      | :sg:`fblits(blit_sequence=((source, dest), ...), special_flags=0) -> None`
+
+      This method takes a sequence of tuples (source, dest) as input, where source is a Surface
+      object and dest is its destination position on this Surface. It draws each source surface
```
Comment (Matiiss) [inline docs/reST/ref/surface.rst:163]: Inconsistencies with capitalizing Surface, it is not capitalized here at the end of the line

File: docs/reST/ref/surface.rst [hunk 6]
```diff
@@ -154,6 +154,30 @@
...
+      fully(meaning that unlike `blit()` you cannot pass an "area" parameter to represent
+      a smaller portion of the source Surface to draw) on this Surface with the same blending
+      mode specified by special_flags. The sequence must have at least one (source, dest) pair.
+
+      :param blit_sequence: a sequence of (source, dest)
+      :param special_flags: the flag(s) representing the blend mode used for each surface
+
+      :returns: always returns ``None``
+      :rtype: None
```
Comment (Matiiss) [inline docs/reST/ref/surface.rst:172]: Are both of these needed? Also for consistency rtype should also be in code quotes I guess or just remove the return type and just do ":returns: ``None``"? Just not sure if that "always" conditional is that necessary, or just leave ":rtype: ``None``"? It just feels a bit redundant IMO

File: test/blit_test.py [hunk 7]
```diff
@@ -128,6 +128,73 @@ def blits(blit_list):
...
+        for i in range(NUM_SURFS):
+            color = (i * 1, i * 1, i * 1)
+            self.assertEqual(dst.get_at((i * 10, 0)), color)
+            self.assertEqual(dst.get_at(((i * 10) + 5, 5)), color)
+
+        t0 = time()
+        results = dst.fblits(blit_list, 0)
+        t1 = time()
+        if PRINT_TIMING:
+            print(f"Surface.fblits: {t1 - t0}")
```
Comment (Matiiss) [inline test/blit_test.py:189]: Oh, I suppose this was meant to be "Surface.fblits special_flags=0"?

**Review by Matiiss — CHANGES_REQUESTED** [review]
Some minor docs stuff pretty much

**Review by Matiiss — APPROVED** [review]
LGTM, I ran the tests with prints and these single runs seemed a bit weird since python blits were almost always faster than fblits but I ran those functions multiple times in a loop in between the timers and there was about 33% of a speedup so that's nice 👍
